src/sqlfluff/rules/L008.py

```diff
@@ -34,45 +36,75 @@ class Rule_L008(BaseRule):
         WHERE a IN ('plop',•'zoo')
     """
 
-    def _eval(self, context: RuleContext) -> Optional[LintResult]:
-        """Commas should be followed by a single whitespace unless followed by a comment.
+    def _get_subsequent_whitespace(
+        self,
+        context,
+    ) -> Tuple[Optional[BaseSegment], Optional[BaseSegment]]:
+        """Search forwards through the raw segments for subsequent whitespace.
 
-        This is a slightly odd one, because we'll almost always evaluate from a point a few places
-        after the problem site. NB: We need at least two segments behind us for this to work.
+        Return a tuple of both the trailing whitespace segment and the
+        first non-whitespace segment discovered.
         """
-        if len(context.raw_stack) < 1:
-            return None
+        subsequent_whitespace = None
+        # Get all raw segments and find position of the current comma within the list.
+        file_segment = context.parent_stack[0]
+        # Raw stack is appropriate as the only segments we can care about are
+        # comma, whitespace, newline, and comment, which are all raw.
+        # Using the raw_segments allows us to account for possible unexpected
+        # parse tree structures resulting from other rule fixes.
```

Comment:
It sounds like you may be seeing an occurrence of #1304. I would really like to get all these rule bugs fixed (I suspect many of them involve misuse of `LintFix` type `edit`) and add a linter check to prevent such bugs in the future.

Other related issues:
* #1304: Another apparent rule bug
* #1012: Vague idea about a general way of detecting all such issues

Reply:
@barrywhart so I think a lot of these issues aren't necessarily an `edit` thing, you would get a similar issue with `create_before` or `create_after` as they all work in a similar way. 

Reading through the `apply_fixes` logic (https://github.com/sqlfluff/sqlfluff/blob/0624a848057dcbf020dd07fe3999605e5f2f973e/src/sqlfluff/core/parser/segments/base.py#L938), a fix is applied by:
- Search down the parse tree, at each level appending each segment you find to segment buffer.
- If you find the anchor segment (of the fix not the result) then either
edit: add the new segments to the buffer, but not the anchor.
create_before: add the new segment to the buffer and then the anchor segment
create_after: add the anchor segment to the buffer and then add the new segment
- then subsequent segments are added to the buffer.

The key thing to note here is that all fixes segments are inserted at the SAME level as the fix anchor_segment. The reason we seem to get a lot of issue like the one above is some logic (in this case I think it's L016) finds a segment to analyse and determine if a fix needs to be applied and then naively just applies the fix at the same tree level even if not appropriate.

So in this case L016 probably determines that it can split a line by looking at the first raw segment after a space, which is the function_name_identifier, and then inserts the newline and whitespace before that. Something like `LintFix("create_before", <function_name_identifier_segment>, [NewlineSegment(), WhitespaceSegment()])`, when in fact from the parse tree the appropriate fix is actually `LintFix("create_before", <expression_segment>, [NewlineSegment(), WhitespaceSegment()])`.
So there needs to be bit of extra logic/consideration to ensure the fix is actually placed in the correct place in the parse tree rather than to just get the output appearing correct in the simple case. That being said it is hard sometimes to implement.

A good example of this is the recent PR we did for #1979, in which we detect the linting error at the final code segment of the tree but apply the fix at the FileSegment in order to get the correct parse structure.

Reply:
^That's basically the cause of these bugs, however, given that it's sometimes hard to create logic to determine the correct fix placement (especially when the parse grammar itself can be inconsistent) I do think that #1012 is an interesting idea.

If you were, for example, to take the messed up post-fix parse tree structure shown above and re-parse the updated `<file_segment>.raw` to then supply to subsequent fix-rounds you would maintain the fixes whilst giving subsequent rules the correct parse tree. (guess it would be between loops in `lint_fix_parsed`)

It might not be the most efficient but could be more practical given that we are unlikely to have completely well defined grammar anytime soon?

Reply:
I have also been wondering if we could run the relevant `match()` functions to detect and complain if a rule does something invalid to the parse tree. I think I need to look at some of the bugs more closely to decide if it's reasonable to flag/fix them.

Re-parsing is an alternative -- it should definitely work, but it feels like cheating, and in order to be 100% sure, we'd potentially need to re-parse after every fix (or every rule run)? It seems like it'd be useful to know which rules have these issues so we only re-parse after running a problematic rule.